File: backend/src/entities/AuditLog.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, Index } from 'typeorm';

@Entity('audit_logs')
@Index(['branchId', 'createdAt'])
@Index(['action', 'createdAt'])
export class AuditLog {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Column({ type: 'uuid', nullable: true })
  userId?: string | null;

  @Column({ type: 'varchar', length: 50, nullable: true })
  username?: string | null;

  @Column({ type: 'varchar', length: 50, nullable: true })
  userRole?: string | null;

  @Column({ type: 'uuid', nullable: true })
  branchId?: string | null;

  @Column({ type: 'varchar', length: 50 })
  action!: string; // 'LOGIN', 'CREATE_SALE', 'VOID_SALE', 'RECORD_PAYMENT', 'VOID_PAYMENT', 'WAIVE_INSTALLMENTS', 'USER_CREATE', etc.

  @Column({ type: 'varchar', length: 50, nullable: true })
  targetEntity?: string | null;

  @Column({ type: 'varchar', length: 100, nullable: true })
  targetId?: string | null;

  @Column({ type: 'text', nullable: true })
  details?: string | null;

  @Column({ type: 'varchar', length: 50, nullable: true })
  ipAddress?: string | null;

  @Column({ type: 'varchar', length: 255, nullable: true })
  userAgent?: string | null;

  @CreateDateColumn()
  createdAt!: Date;
}

File: backend-java/src/main/java/com/murabha/cloud/entity/AuditLog.java
package com.murabha.cloud.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "audit_logs", indexes = {
    @Index(name = "idx_audit_user_id", columnList = "user_id"),
    @Index(name = "idx_audit_branch_id", columnList = "branch_id"),
    @Index(name = "idx_audit_action", columnList = "action"),
    @Index(name = "idx_audit_created_at", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id")
    private UUID userId;

    @Column(length = 64)
    private String username;

    @Column(length = 32)
    private String role;

    @Column(name = "branch_id")
    private UUID branchId;

    @Column(nullable = false, length = 64)
    private String action;

    @Column(name = "target_entity", length = 64)
    private String targetEntity;

    @Column(name = "target_id", length = 64)
    private String targetId;

    @Column(columnDefinition = "TEXT")
    private String details;

    @Column(name = "ip_address", length = 64)
    private String ipAddress;

    @Column(name = "user_agent", length = 255)
    private String userAgent;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
}